src/util/log/fd_log.c

```diff
@@ -495,39 +498,39 @@ void fd_log_level_stderr_set ( int level ) { FD_VOLATILE( fd_log_private_level_s
 void fd_log_level_flush_set  ( int level ) { FD_VOLATILE( fd_log_private_level_flush   ) = level; }
 void fd_log_level_core_set   ( int level ) { FD_VOLATILE( fd_log_private_level_core    ) = level; }
 
-/* Total size of the log buffer (fd_log_private_log_msg) */
+/* Buffer size used for vsnprintf calls (this is also one more than the
+   maximum size that this can passed to fd_io_write) */
 
 #define FD_LOG_BUF_SZ (4UL*4096UL)
 
-/* Log buffer, used by fd_log_private_0 and fd_log_private_hexdump_msg */
+/* Lock to used by fd_log_private_fprintf_0 to sequence calls writes
+   between different _processes_ that share the same fd. */
 
-static FD_TLS char fd_log_private_log_msg[ FD_LOG_BUF_SZ ];
-static FD_TLS char fd_log_private_log_msg0[ FD_LOG_BUF_SZ ];
+static int * fd_log_private_shared_lock; /* NULL outside boot/halt, init at boot */
 
-char const *
-fd_log_private_0( char const * fmt, ... ) {
-  va_list ap;
-  va_start( ap, fmt );
-  int len = vsnprintf( fd_log_private_log_msg, FD_LOG_BUF_SZ, fmt, ap );
-  if( len<0    ) len = 0;    /* cmov */
-  if( len>(int)FD_LOG_BUF_SZ-1 ) len = FD_LOG_BUF_SZ-1; /* cmov */
-  fd_log_private_log_msg[ len ] = '\0';
-  va_end( ap );
-  return fd_log_private_log_msg;
-}
+static int fd_log_private_shared_lock_local[1] __attribute__((aligned(128))); /* location of lock if boot mmap fails */
 
-/* Lock to sequence writes between different threads */
+void
+fd_log_private_fprintf_0( int          fd,
+                          char const * fmt, ... ) {
 
-static int * fd_log_private_shared_lock = NULL;
+  /* TODO:
+     - Consider moving to util/io as fd_io_printf or renaming to
```

Comment:
Probably a good idea, I already needed to call this "private" from some of the code in run.c that does shutdown logging inside signal handlers.